src/sinks/util/tls.rs

```diff
@@ -0,0 +1,192 @@
+    DerExportError { source: openssl::error::ErrorStack },
+    #[snafu(display("Could not parse certificate in {:?}: {}", filename, source))]
+    CertificateParseError {
+        filename: PathBuf,
+        source: native_tls::Error,
+    },
+    #[snafu(display("Must specify both TLS key_file and crt_file"))]
+    MissingCrtKeyFile,
+    #[snafu(display("Could not parse X509 certificate in {:?}: {}", filename, source))]
+    X509ParseError {
+        filename: PathBuf,
+        source: openssl::error::ErrorStack,
+    },
+    #[snafu(display("Could not parse private key in {:?}: {}", filename, source))]
+    PrivateKeyParseError {
+        filename: PathBuf,
+        source: openssl::error::ErrorStack,
+    },
+    #[snafu(display("Could not build PKCS#12 archive for identity: {}", source))]
+    Pkcs12Error { source: openssl::error::ErrorStack },
+}
+
+/// Standard TLS connector options
+#[derive(Debug, Clone, Derivative, Deserialize, Serialize)]
+#[derivative(Default)]
+pub struct TlsOptions {
+    #[derivative(Default(value = "true"))]
+    pub verify_certificate: bool,
+    pub ca_path: Option<PathBuf>,
+    pub crt_path: Option<PathBuf>,
+    pub key_path: Option<PathBuf>,
+    pub key_pass: Option<String>,
+}
+
+/// Directly usable settings for TLS connectors
+#[derive(Derivative)]
+#[derivative(Default)]
+pub struct TlsSettings {
+    #[derivative(Default(value = "true"))]
+    verify_certificate: bool,
+    authority: Option<Certificate>,
+    identity: Option<Identity>,
+}
+
+impl TryFrom<&TlsOptions> for TlsSettings {
+    type Error = crate::Error;
+    fn try_from(options: &TlsOptions) -> Result<Self, Self::Error> {
+        if options.crt_path.is_some() != options.key_path.is_some() {
+            return Err(TlsError::MissingCrtKeyFile.into());
+        }
+
+        let authority = match options.ca_path {
+            None => None,
+            Some(ref path) => Some(load_certificate(path)?),
+        };
+
+        let identity = match options.crt_path {
+            None => None,
+            Some(ref crt_path) => {
+                let identity = load_build_pkcs12(
+                    options.key_path.as_ref().unwrap(),
+                    &options.key_pass,
+                    crt_path,
+                )?;
+                Some(
+                    Identity::from_pkcs12(&identity.to_der().context(DerExportError)?, "")
+                        .context(TlsIdentityError)?,
+                )
+            }
+        };
+
+        Ok(Self {
+            verify_certificate: options.verify_certificate,
+            authority,
+            identity,
+        })
+    }
+}
+
+pub trait TlsConnectorExt {
+    fn use_tls_settings(&mut self, settings: TlsSettings) -> &mut Self;
+}
+
+impl TlsConnectorExt for TlsConnectorBuilder {
+    fn use_tls_settings(&mut self, settings: TlsSettings) -> &mut Self {
+        self.danger_accept_invalid_certs(!settings.verify_certificate);
```

Comment:
might be nice to add a warning here that this is being enabled.

Reply:
It is optionally enabled by setting `verify_certificate` to `false`, and there is a warning in the documentation for that setting (in the `http` sink at least, where it is visible).

... but now that I look at it, it should be an `Option<bool>` which defaults to `true`. I'll fix that part.

Reply:
Oh, and I see the `http` sink does output a warning. I'll look into consolidating the two.

Reply:
Done